internal/pkg/vm/internal/vmcontext/vmcontex2.go

```
@@ -0,0 +1,543 @@
+		rewardMessage := makeBlockRewardMessage(blk.Miner, minerPenaltyTotal)
+		vm.applyImplicitMessage(rewardMessage)
+	}
+
+	// cron tick
+	firstMiner := blocks[0].Miner
+	cronMessage := makeCronTickMessage(firstMiner)
+	vm.applyImplicitMessage(cronMessage)
+
+	// commit state
+	// flush all objects out
+	if err := vm.store.Flush(); err != nil {
+		// Review: what do we want to do with this errors?
+		//         we can change the return to be ([]Receipt, error) although its unclear what to do about it as a caller
+		//         we can just crash, or let the caller retry a few times and then crash
+		panic(err)
+	}
+	// commit new actor state
+	if err := vm.state.Commit(context.Background()); err != nil {
+		// Review: what do we want to do with this errors?
+		//         similarly, what can the caller do?
+		panic(err)
+	}
+	// TODO: update state root (issue: #3718)
+
+	return receipts
+}
+
+// applyImplicitMessage applies messages automatically generated by the vm itself.
+//
+// This messages do not consume client gas are do not fail.
+func (vm *VM) applyImplicitMessage(imsg internalMessage) interface{} {
+	// implicit messages gas is tracked separatly and not paid by the miner
+	gasTank := gas.NewTracker(gas.SystemGasLimit)
+
+	// the execution of the implicit messages is simpler than full external/actor-actor messages
+	// execution:
+	// 1. load from actor
+	// 2. increment seqnumber
+	// 3. build new context
+	// 4. invoke message
+
+	// 1. load from actor
+	fromActor, err := vm.state.GetActor(context.Background(), imsg.from)
+	if fromActor == nil || err != nil {
+		panic("implicit message `from` actor not found")
+	}
+
+	// 2. increment seq number
+	fromActor.IncrementSeqNum()
+
+	// 3. build context
+	ctx := newInvocationContext(vm, imsg, fromActor, &gasTank)
+
+	// 4. invoke message
+	// Note: any errors raised by the invocation will not be trapped
+	// 		 errors are not expected by the spec on implicit messages, this messages must work
+	return ctx.invoke()
```

I think this should propagate an explicit error